### PR TITLE
refactor(components): extract shared :root token-name introspection for tests

### LINKED ISSUE #771: Triplicated `:root` token-extraction logic across two packages
## Summary

Three files independently hand-roll the same `tokens-base.css` `:root` extraction, with a **byte-identical** regex:

| File | Line |
| --- | --- |
| `packages/components/src/styles/tokens.test.ts` | 17 |
| `packages/components/src/styles/tokens-doc-drift.test.ts` | 25 |
| `packages/playground/src/shell-app/color-token-registry.test.ts` | 24 |

```js
css.match(/^\s*:root\s*\{([\s\S]*?)\n\}/m)
```

Two of them (`tokens-doc-drift`, `color-token-registry`) go further and duplicate the *entire* helper — same regex, same `/\*...\*/` comment strip, same `/(--cinder-[a-z0-9-]+)\s*:/g` name pattern — across a package boundary.

## Why it matters

That regex is quietly formatting-coupled. It requires:

- the `:root` block to be the **first** `^\s*:root\s*{` in the file, and
- its closing brace to sit at **column 0** (`\n\}`), since the match is non-greedy.

Reindent `tokens-base.css`, or introduce an earlier `:root` block, and all three break — independently, in three different test suites, with three different error messages. `tokens-doc-drift.test.ts` and `color-token-registry.test.ts` both `throw` on no-match, so the failure is at least loud; but it is three separate fixes for one root cause.

This is a real coupling: `tokens-base.css` already contains `:root[data-theme='dark']` (line 466) and bare `[data-theme='dark']` (474) blocks. Today those don't match the pattern (`[` follows `:root`), so it works — but the margin is one selector edit wide.

## Suggested fix

Extract a single shared helper (`readRootTokenNames(css)`), export it from the components package, and have all three call it. One regex, one place, one test.

## Explicitly not in scope

`tokens.test.ts:58` uses a **deliberately different** name pattern (`/--([a-zA-Z0-9_-]+)\s*:/g`) because it validates per-component token files and must catch `--_cinder-*` privates and malformed names. That divergence is correct and should stay.

## Notes

Surfaced while evaluating cinder against the [DTCG 2025.10 spec](https://www.designtokens.org/tr/2025.10/). Related but separate: #770.

### PR DESCRIPTION
**Pull request by stevekinney**
## Summary

Closes #771.

Three test suites hand-rolled the same byte-identical regex to pull the `:root { ... }` block out of `tokens-base.css`:

- `packages/components/src/styles/tokens.test.ts`
- `packages/components/src/styles/tokens-doc-drift.test.ts`
- `packages/playground/src/shell-app/color-token-registry.test.ts`

The latter two duplicated the *entire* helper (regex + comment strip + name pattern) across the `components`/`playground` package boundary. The regex was formatting-coupled — it required the `:root` block to be the first bare `:root {` in the file, with its closing brace at column 0.

## What changed

- Added `packages/components/src/test/token-introspection.ts`, exporting:
  - `extractRootBlock(css)` — returns the raw source text of the first top-level, unscoped `:root { ... }` block.
  - `readRootTokenNames(css)` — returns the `Set<string>` of `--cinder-*` custom property names declared in that block.
- Both are implemented with **PostCSS** (already a `components` dependency, already the house style for CSS test helpers — see `src/test/css.ts`) instead of a hand-rolled regex/brace scan. This hardens the formatting coupling the issue called out:
  - Reindentation is a non-issue — PostCSS parses the AST, not source columns.
  - Scoped variants like `:root[data-theme='dark']` are never matched — PostCSS compares the full rule selector, and an attribute-qualified selector is a different string than `:root`.
  - A `:root` block nested inside `@media (prefers-reduced-motion: reduce) { ... }` is skipped — only the top-level (non-nested) rule is considered.
  - Comments can never be mistaken for declarations — PostCSS parses them as distinct nodes that are simply not walked.
- All three call sites now import from the shared helper and their local duplicate implementations are gone.
- `packages/components/src/test/token-introspection.test.ts` — new tests proving the hardened behavior: reindented block, a preceding `:root[data-theme]` block, a nested `@media` block, and comment-stripping.
- `packages/components/scripts/check-consumer-boundaries.ts` — the boundary guard previously carved out a single hardcoded exception (`src/test/happy-dom.ts`) so playground *test* files could reach into `components/src/test/` for shared test infrastructure. **Final shape (after review): an explicit allowlist**, not a directory- or filename-pattern-based exemption. Review (Copilot + Codex, across several rounds) found that any pattern-based exemption reopens a private-source reach-in: `src/test/fixtures/**` Svelte fixtures import private component source directly, `src/test/*.test.ts` files do the same to test it, and `src/test/index.ts`/`server-render.ts` export helpers that accept an arbitrary `sourcePath` string at *runtime* and dynamically import it — invisible to a static import-specifier check no matter how the filename pattern is tuned. The guard now checks an explicit `Set` containing exactly the two helpers playground tests use: `packages/components/src/test/happy-dom.ts` and `packages/components/src/test/token-introspection.ts`. Adding a future helper requires a conscious edit to that allowlist. Regression tests cover all of the above (fixtures, `.test.ts`, `index.ts`/`server-render.ts`) staying rejected.

## Why not a `package.json` export

`token-introspection.ts` is test-only infrastructure, not a runtime API. The `styles/guard`-style subpath-export route would require wiring it into the hand-authored reserved-key machinery in `scripts/generate-exports.ts` (`RESERVED_KEYS`, a canonical entry function, check-mode drift detection, `files` array) for something no runtime consumer needs. The `src/test/*` relative-import allowance already exists for exactly this shape of problem (see `happy-dom.ts`), so this follows that precedent instead — now as an explicit allowlist rather than a pattern (see above).

## Changeset

None. Every touched file is a `*.test.ts`, `scripts/check-consumer-boundaries.ts`, or `src/test/token-introspection.ts` — none are in `packages/components/package.json#files`, so nothing here ships in the published package.

## Test plan

- [x] `bun run --filter=@lostgradient/cinder typecheck` — 0 errors
- [x] `bun run --filter=@cinder/playground typecheck` — 0 errors
- [x] `bun run --filter=@lostgradient/cinder test` (full suite, real DOM conditions) — 7210 pass / 0 fail
- [x] `bun run --filter=@cinder/playground test` (full suite) — 895 pass / 0 fail
- [x] `bun run --filter=@lostgradient/cinder lint:invariants` (includes `check:consumer-boundaries` against the real tree) — OK
- [x] Pre-commit hook (lint-staged + full typecheck for both touched packages) — passed

### CODE REVIEW

File: packages/components/src/test/token-introspection.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
+/**
+ * Shared `tokens-base.css` `:root` introspection for tests. Internal — not
+ * part of the public package surface (see `src/test/README` conventions:
+ * these helpers live outside `package.json#exports` and are reached via the
+ * `packages/components/src/test/*` relative-import allowance that
+ * `check-consumer-boundaries.ts` carves out for test files).
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/test/token-introspection.ts]: Good catch — that referenced a `src/test/README` that never existed (I was thinking of the convention, not an actual file). Fixed by rewriting the sentence to point at the two real files that already establish this pattern (`src/test/happy-dom.ts`, `src/test/css.ts`) instead of a nonexistent doc.

File: packages/components/src/test/token-introspection.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+  root.walkRules(':root', (rule) => {
+    if (found || rule.parent?.type !== 'root') return;
+    found = true;
+
+    for (const node of rule.nodes) {
+      if (node.type === 'decl' && node.prop.startsWith('--cinder-')) {
+        names.add(node.prop);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/test/token-introspection.ts:68]: Reasoned disagreement here — I don't think this can throw in practice, and I'd rather not add a defensive `?? []` that masks what the type system is already telling us.

`Container.nodes` is indeed typed as `Child[] | undefined` in `postcss/lib/container.d.ts` (line 63) — but `Rule` narrows that. `postcss/lib/rule.d.ts` line 82 declares:

```ts
nodes: NonNullable<Container['nodes']>
```

So for any `Rule` obtained from `parse()` (which is what `walkRules` always hands the callback), TypeScript already guarantees `rule.nodes` is non-undefined — that's why `bun run --filter=@lostgradient/cinder typecheck` is clean without an `?? []`. I also verified empirically: parsing `:root {}` and reading `rule.nodes` off the matched rule returns `[]`, not `undefined` — PostCSS's `Container` constructor always initializes `nodes` to an array, empty rule or not. The `Child[] | undefined` on the base `Container` type exists for other container kinds (or hand-constructed AST nodes), not for a rule PostCSS itself produced from parsing.

Given that, adding `rule.nodes ?? []` here would be dead code that contradicts the type system, which is exactly the kind of unnecessary defensive pattern I'd rather avoid. Happy to add it if there's a PostCSS version/edge case I'm missing — but as far as the installed `postcss@8.5.x` types and a live test show, this branch isn't reachable.

File: packages/components/scripts/check-consumer-boundaries.ts [hunk 3]
```diff
@@ -48,7 +48,7 @@ export function findConsumerBoundaryViolations(
       importTarget === 'packages/components/src/index.ts' ||
       ((normalizedFilePath.endsWith('.test.ts') ||
         normalizedFilePath === 'packages/playground/scripts/preload.ts') &&
-        importTarget === 'packages/components/src/test/happy-dom.ts')
+        importTarget.startsWith('packages/components/src/test/'))
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/scripts/check-consumer-boundaries.ts]: Good find, thanks — that was a real gap. Narrowed the exemption from "anything under src/test/" to a regex requiring a flat `.ts` file directly under `src/test/` (no nested path): `^packages/components/src/test/[^/]+\.ts$`. That still covers the two real consumers (`happy-dom.ts`, `token-introspection.ts`) but now rejects `src/test/fixtures/*.svelte` exactly as you described, since those fixtures import private component source directly. Added a regression test (`rejects src/test/fixtures/** even from tests`) asserting the fixture import you flagged now produces a violation, and re-ran `bun run check:consumer-boundaries` against the real tree — still OK.

[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/scripts/check-consumer-boundaries.ts]: Fixed in 356fe6f3 — good catch, the exemption was too broad.

It is now scoped to flat `.ts` files directly under `src/test/` (`/^packages\/components\/src\/test\/[^/]+\.ts$/`), which deliberately excludes `src/test/fixtures/**`. Those Svelte fixtures import private component source themselves, so exempting them would have let a playground test transitively reach private component source through a fixture — exactly the reach-in this guard exists to prevent. The reasoning is recorded in a comment above the pattern, and `check-consumer-boundaries.test.ts` gained coverage for the nested-fixture case so the narrower scope cannot silently widen again.

File: packages/components/scripts/check-consumer-boundaries.ts [hunk 4]
```diff
@@ -6,6 +6,15 @@ const playgroundRoots = ['packages/playground/src', 'packages/playground/scripts
...
   /(?:from\s*|(?:import|require|import\.meta\.glob)\s*\(\s*(?:\/\*[\s\S]*?\*\/\s*)*|import\s+)(['"`])((?:\\.|(?!\1)[\s\S])*?)\1/g;
+// Shared test helper modules such as `src/test/happy-dom.ts` and
+// `src/test/token-introspection.ts` — flat `.ts` files directly under
+// `src/test/`, never a nested path. Deliberately excludes `src/test/fixtures/**`:
+// those Svelte fixtures import PRIVATE component source themselves (e.g. an
+// `access-gate-*-fixture.svelte` importing `../../components/access-gate/access-gate.svelte`),
+// so allowing playground tests to reach *those* would let a playground test
+// transitively depend on private component source through the fixture —
+// exactly the reach-in this guard exists to prevent.
+const sharedTestHelperPattern = /^packages\/components\/src\/test\/[^/]+\.ts$/;
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/scripts/check-consumer-boundaries.ts]: Right, this was the exact same class of gap you and Codex both caught (Codex flagged it via hydrate.test.ts specifically). Fixed: the pattern now has a negative lookahead excluding .test.ts and .d.ts (`^packages/components/src/test/(?!.*\.(?:test|d)\.ts$)[^/]+\.ts$`). Added a regression test asserting both token-introspection.test.ts and hydrate.test.ts stay rejected.

File: packages/components/scripts/check-consumer-boundaries.test.ts [hunk 5]
```diff
@@ -37,6 +37,31 @@ describe('consumer boundary guard', () => {
...
+  test('allows a flat shared test helper .ts file directly under src/test/ only from tests', () => {
+    const source =
+      "import { readRootTokenNames } from '../../components/src/test/token-introspection.ts';";
+    expect(findConsumerBoundaryViolations(source, 'packages/playground/src/app.test.ts')).toEqual(
+      [],
+    );
+    expect(findConsumerBoundaryViolations(source, 'packages/playground/src/app.ts')).toHaveLength(
+      1,
+    );
+  });
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/scripts/check-consumer-boundaries.test.ts]: Added — new test `rejects src/test/*.test.ts even from tests — co-located tests import private component source` asserts both `token-introspection.test.ts` and `hydrate.test.ts` (the file Codex pointed at in the sibling thread) still produce a violation from a playground `*.test.ts` file.

File: packages/components/scripts/check-consumer-boundaries.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/scripts/check-consumer-boundaries.ts]: Confirmed and fixed — thanks for the concrete example. `hydrate.test.ts:21` importing `../components/input/input.svelte` was exactly the shape of transitive reach-in I hadn't excluded. The pattern now excludes `.test.ts` (and `.d.ts`) via a negative lookahead, so `import '../../../components/src/test/hydrate.test.ts'` from a playground test now produces a violation — added as a regression test alongside the `token-introspection.test.ts` case Copilot raised in the sibling thread.

File: packages/components/scripts/check-consumer-boundaries.ts [hunk 6]
```diff
@@ -6,6 +6,20 @@ const playgroundRoots = ['packages/playground/src', 'packages/playground/scripts
...
+// test transitively depend on PRIVATE component source through an
+// intermediary — exactly the reach-in this guard exists to prevent:
+//   - `src/test/fixtures/**`: Svelte fixtures that import private component
+//     source directly (e.g. an `access-gate-*-fixture.svelte` importing
+//     `../../components/access-gate/access-gate.svelte`).
+//   - `src/test/*.test.ts`: component tests co-located in this directory
+//     (e.g. `hydrate.test.ts`) that import private component source to test
+//     it — a shared-helper exemption is not meant to cover test files.
+const sharedTestHelperPattern =
+  /^packages\/components\/src\/test\/(?!.*\.(?:test|d)\.ts$)[^/]+\.ts$/;
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/scripts/check-consumer-boundaries.ts]: You're right, and this is the real fix — thank you for pushing on this through two rounds. A filename pattern can never account for a helper that accepts an arbitrary path as a runtime argument (`renderThenHydrate`/`renderToServerHtml`'s `sourcePath`) rather than expressing its dependency as a static import specifier, so no amount of tuning the pattern (excluding `.test.ts`, excluding `fixtures/`) would have closed this off.

Replaced the pattern entirely with an explicit `Set` of the two helpers playground tests actually use (`happy-dom.ts`, `token-introspection.ts`). `index.ts` and `server-render.ts` are no longer reachable at all. Added a regression test asserting both are rejected. Confirmed `bun run check:consumer-boundaries` still passes against the real tree (only the two allowlisted files are ever imported from playground tests today).

[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/scripts/check-consumer-boundaries.ts]: Fixed — you were right that the pattern still had a hole, and it is now an explicit allowlist rather than a filename regex.

Only two helpers are reachable: `src/test/happy-dom.ts` and `src/test/token-introspection.ts`. `index.ts` and `server-render.ts` are deliberately excluded for exactly the reason you gave — the barrel re-exports `renderThenHydrate`/`renderToServerHtml`, which take an arbitrary `sourcePath` and dynamically import it, so allowing the FILE cannot rule out a playground test passing a private `src/components/**/*.svelte` path through as data rather than as a static import specifier.

The comment above the allowlist records all three bypasses the pattern kept reopening (nested fixtures, co-located `*.test.ts`, and this barrel case), so the next person to widen it has to do so consciously. `check-consumer-boundaries.test.ts` is green at 22/22.

File: packages/components/scripts/check-consumer-boundaries.ts [hunk 7]
```diff
@@ -6,6 +6,34 @@ const playgroundRoots = ['packages/playground/src', 'packages/playground/scripts
...
 const sourceImportPattern =
   /(?:from\s*|(?:import|require|import\.meta\.glob)\s*\(\s*(?:\/\*[\s\S]*?\*\/\s*)*|import\s+)(['"`])((?:\\.|(?!\1)[\s\S])*?)\1/g;
+// Explicit allowlist of shared test helpers that `packages/playground` test
+// files (and `scripts/preload.ts`) may reach into directly, alongside the
+// `happy-dom.ts` precedent this follows.
+//
+// This started as a filename PATTERN (any flat, non-`.test.ts` file directly
+// under `src/test/`) but that kept reopening the same class of bypass as new
+// cases surfaced during review:
+//   - `src/test/fixtures/**` — Svelte fixtures that import private component
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/scripts/check-consumer-boundaries.ts:16]: Good catch — the description was stale from before the allowlist rewrite. Updated the "What changed" and "Why not a package.json export" sections to describe the actual final implementation (explicit Set of two files) instead of the earlier directory-wide exemption.
